les/utils/limiter.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// Limiter protects a network request serving mechanism from denial-of-service attacks.
+// It limits the total amount of resources used for serving requests while ensuring that
+// the most valuable connections always have a reasonable chance of being served.
+type Limiter struct {
+	lock sync.Mutex
+	cond *sync.Cond
+	quit bool
+
+	nodes                                map[enode.ID]*nodeQueue
+	addresses                            map[string]*addressGroup
+	addressSelect, valueSelect           *WeightedRandomSelect
+	maxValue                             float64
+	maxWeight, sumWeight, sumWeightLimit uint
+	selectedByValue                      bool
+}
+
+// nodeQueue represents queued requests coming from a single node ID
+type nodeQueue struct {
+	queue                   []request
+	id                      enode.ID
+	address                 string
+	value                   float64
+	flatWeight, valueWeight uint64 // current selection weights in the address/value selectors
+	nextWeight              uint   // next request weight
```

Comment:
Perhaps: `flatSelection`, and `valueSelection` would be better?

Reply:
Calling the selection weights "selection" also does not really describe what those numbers actually are (sounds more like it has something to do with the actual selected item).
Since we've already had the "weighted random selector" for a long time, I decided to keep the term "weight" for selection weights and renamed request weight to request cost (since we already use the term "cost" for LES requests).